File: src/python/ko_pii/patterns/card.py
"""신용카드 / 체크카드 번호 (Card Number) detection.

Detection requires the Luhn (mod-10) checksum to pass + IIN/BIN 첫 자리가
실제 카드 브랜드의 화이트리스트에 있어야 한다 (ISO/IEC 7812).

지원 BIN 첫 자리:
  3 - American Express, JCB, Diners Club
  4 - Visa
  5 - Mastercard
  6 - Discover / UnionPay
  9 - 한국 국내 전용 카드 (BC, 일부 백화점카드 등)
  2 - Mastercard 신규 BIN (2221~2720), 일부 한국 카드

거부: 0, 1, 7, 8 — 현행 카드 브랜드에 할당되지 않음.

Luhn 만으로는 "1111-2222-3333-4444" 같이 trivial 한 반복 숫자가 우연히
통과하여 FP 가 됨. 따라서 첫 자리 화이트리스트로 한 단계 더 거름.

Legal basis: 개인정보보호법 제2조; 여신전문금융업법.
"""
from __future__ import annotations

import re
from typing import Iterator

from ko_pii.checksum.luhn import is_valid as is_valid_luhn
from ko_pii.core.types import DetectionResult, RiskLevel

LABEL = "CARD"
LEGAL_BASIS = "개인정보보호법 제2조; 여신전문금융업법"
CATEGORY = "일반개인정보"

_VALID_BIN_FIRST_DIGITS: frozenset[str] = frozenset({"2", "3", "4", "5", "6", "9"})

_PATTERN = re.compile(
    r"(?<![0-9])"
    r"(?:"
    r"[0-9]{4}[-. /]\s?[0-9]{4}[-. /]\s?[0-9]{4}[-. /]\s?[0-9]{1,7}"
    r"|"
    r"[0-9]{13,19}"
    r")"
    r"(?![0-9])"
)


def _brand_for(first_digit: str) -> str:
    return {
        "3": "amex_or_jcb_or_diners",
        "4": "visa",
        "5": "mastercard",
        "6": "discover_or_unionpay",
        "9": "korea_domestic",
        "2": "mastercard_new_or_misc",
    }.get(first_digit, "unknown")


def detect(text: str) -> Iterator[DetectionResult]:
    for m in _PATTERN.finditer(text):
        raw = m.group(0)
        digits = re.sub(r"\D", "", raw)
        if not (13 <= len(digits) <= 19):
            continue
        if digits[0] not in _VALID_BIN_FIRST_DIGITS:
            continue
        # 길이-브랜드 일관성: 존재하지 않는 조합을 거른다. 13자리 카드는 구형 Visa(4)
        # 뿐이고 15자리는 Amex(34/37)뿐 — 이 규칙은 실 카드 recall 을 해치지 않으면서
        # EAN-13 바코드(3·5 로 시작)나 15자리 IMEI(35…)의 Luhn-우연 통과 FP 를 막는다.
        if len(digits) == 13 and digits[0] != "4":
            continue
        if len(digits) == 15 and digits[:2] not in {"34", "37"}:
            continue
        if not is_valid_luhn(digits):
            continue
        yield DetectionResult(
            label=LABEL,
            text=raw,
            start=m.start(),
            end=m.end(),
            risk_level=RiskLevel.HIGH,
            confidence=1.0,
            evidence=[
                "pattern:card",
                "checksum:luhn_valid",
                f"brand_hint:{_brand_for(digits[0])}",
            ],
            legal_basis=LEGAL_BASIS,
            extra={
                "digits": digits,
                "length": len(digits),
                "brand_hint": _brand_for(digits[0]),
                "category": CATEGORY,
            },
        )

File: src/ts/src/patterns/card.ts
/** 신용카드 / 체크카드 번호 (Card Number) detection.
 *
 * Python 원본: src/ko_pii/patterns/card.py — 1:1 포팅 (골드 마스터 기준).
 *
 * Luhn (mod-10) 체크섬 통과 + IIN/BIN 첫 자리 화이트리스트 (ISO/IEC 7812).
 * 거부: 0, 1, 7, 8 — 현행 카드 브랜드에 할당되지 않음.
 */

import { isValid as isValidLuhn } from "../checksum/luhn.js";
import { type DetectionResult, makeDetection, RiskLevel } from "../core/types.js";

const LABEL = "CARD";
const LEGAL_BASIS = "개인정보보호법 제2조; 여신전문금융업법";
const CATEGORY = "일반개인정보";

const VALID_BIN_FIRST_DIGITS: ReadonlySet<string> = new Set(["2", "3", "4", "5", "6", "9"]);

const PATTERN =
  /(?<![0-9])(?:[0-9]{4}[-. /]\s?[0-9]{4}[-. /]\s?[0-9]{4}[-. /]\s?[0-9]{1,7}|[0-9]{13,19})(?![0-9])/g;

const BRAND_BY_FIRST_DIGIT: Record<string, string> = {
  "3": "amex_or_jcb_or_diners",
  "4": "visa",
  "5": "mastercard",
  "6": "discover_or_unionpay",
  "9": "korea_domestic",
  "2": "mastercard_new_or_misc",
};

function brandFor(firstDigit: string): string {
  return BRAND_BY_FIRST_DIGIT[firstDigit] ?? "unknown";
}

export function detect(text: string): DetectionResult[] {
  const out: DetectionResult[] = [];
  for (const m of text.matchAll(PATTERN)) {
    if (m.index === undefined) continue;
    const raw = m[0];
    const digits = raw.replace(/\D/g, "");
    if (!(digits.length >= 13 && digits.length <= 19)) continue;
    const first = digits[0];
    if (first === undefined || !VALID_BIN_FIRST_DIGITS.has(first)) continue;
    // 길이-브랜드 일관성: 13자리는 구형 Visa(4), 15자리는 Amex(34/37)만 허용 —
    // EAN-13 바코드/15자리 IMEI 의 Luhn-우연 통과 FP 를 막는다.
    if (digits.length === 13 && first !== "4") continue;
    if (digits.length === 15 && !["34", "37"].includes(digits.slice(0, 2))) continue;
    if (!isValidLuhn(digits)) continue;
    out.push(
      makeDetection({
        label: LABEL,
        text: raw,
        start: m.index,
        end: m.index + raw.length,
        riskLevel: RiskLevel.HIGH,
        confidence: 1.0,
        evidence: ["pattern:card", "checksum:luhn_valid", `brand_hint:${brandFor(first)}`],
        legal_basis: LEGAL_BASIS,
        extra: {
          digits,
          length: digits.length,
          brand_hint: brandFor(first),
          category: CATEGORY,
        },
      }),
    );
  }
  return out;
}
